test-data/unit/check-inference.test

```diff
@@ -3263,3 +3263,68 @@ from typing import Dict, Iterable, Tuple, Union
 def foo(x: Union[Tuple[str, Dict[str, int], str], Iterable[object]]) -> None: ...
 foo(("a", {"a": "b"}, "b"))
 [builtins fixtures/dict.pyi]
+
+[case testUseSupertypeAsInferenceContext]
+# flags: --strict-optional
+from typing import List, Optional
+
+class B:
+    x: List[Optional[int]]
+
+class C(B):
+    x = [1]
+
+reveal_type(C().x)  # N: Revealed type is "builtins.list[Union[builtins.int, None]]"
+[builtins fixtures/list.pyi]
+
+[case testUseSupertypeAsInferenceContextInvalidType]
+from typing import List
+class P:
+    x: List[int]
+class C(P):
+    x = ['a']  # E: List item 0 has incompatible type "str"; expected "int"
+[builtins fixtures/list.pyi]
+
+[case testUseSupertypeAsInferenceContextPartial]
+from typing import List
+
+class A:
+    x: List[str]
+
+class B(A):
```

Comment:
Do we care about cases like 

```python
class A:
    x: List[str]

class B:
    x: List[int]

class C(A, B):
     x = []
```

in the scope of this PR? 
I think that it is already covered by other checks, but do we need to have extra features integration test / error message check?

I think it falls into `# Multiple incompatible candidates, cannot use any of them as context.` branch.

Reply:
Yes, there is a test below (we will give usual "need annotation" error and infer `list[Any]`).